### ISSUE TITLE
Cannot read from built-in variables in custom function in shader

### ISSUE THREAD

**Issue by Bauxitedev** [godotengine/godot#18315]
**Godot version:**
3.0.2 Mono

**OS/device including version:**
KDE Neon
GTX 1060

**Issue description:**
In a shader, you can write custom functions to calculate things. However, you cannot access built-in variables like `TIME` and `VIEWPORT_SIZE` from them, which means you have to pass them in when you call them from `fragment()` or `vertex()`, which is quite annoying and makes the code much less readable.

**Steps to reproduce:**
1. Create new project, add MeshInstance with a ShaderMaterial.
2. Add this shader to it:
```
shader_type spatial;

vec3 my_function()
{
	return vec3(0.5 * sin(TIME * 10.0) + 0.5);
}

void fragment()
{
	ALBEDO = my_function();
}
```
3. I expect the mesh to start flickering. Instead I get an error, saying `TIME` is not defined.

Apparently, you can only use `TIME` from within `fragment()` and `vertex()`.

**Comment by Zireael07** [comment]
Oh yeah, I ran into this too when making the clouds shader. I thought I was doing something wrong so I just avoided TIME in my own function...

**Comment by Zylann** [comment]
I think the reason is, the contents of `fragment` and `vertex` are actually copied into a larger `main()` function in the GLSL shader code. Some variables may not be available because they are local to this function. So if you make a separate function, it makes sense that you can't access them.

However, some of those variables should be available, if they are strictly uniforms that don't have modifications on them before your code gets run, and of course if you don't write to them.

**Comment by Chaosus** [comment]
You can pass TIME as parameter to required function(its inelegant but works as temporary solution)

**Comment by Zireael07** [comment]
Any progress on fixing this?

**Comment by Chaosus** [comment]
Closed as a duplicate of #14904
